takari-builder/src/main/java/io/takari/builder/internal/maven/MavenDependencyResolver.java

```diff
@@ -0,0 +1,308 @@
+  }
+
+  // returns possibly empty collection of resolved dependencies
+  @Override
+  public Map<IArtifactMetadata, Path> getProjectDependencies(boolean transitive,
+      ResolutionScope scope) {
+
+    Set<Artifact> artifacts = resolveDependencies(scope, transitive);
+    Map<IArtifactMetadata, Path> results = new LinkedHashMap<>();
+
+    artifacts.forEach(a -> {
+      results.put(new ArtifactMetadataImpl(a), a.getFile().toPath());
+    });
+
+    return results;
+  }
+
+  private boolean matchesArtifact(Dependency dependency, Artifact a) {
+    return dependency.matchesArtifact(a.getGroupId(), a.getArtifactId(), a.getVersion(),
+        a.getClassifier());
+  }
+
+  private static class ArtifactMetadataImpl implements IArtifactMetadata {
+
+    private final Artifact artifact;
+
+    ArtifactMetadataImpl(Artifact artifact) {
+      this.artifact = artifact;
+    }
+
+    @Override
+    public String getGroupId() {
+      return artifact.getGroupId();
+    }
+
+    @Override
+    public String getArtifactId() {
+      return artifact.getArtifactId();
+    }
+
+    @Override
+    public String getVersion() {
+      return artifact.getVersion();
+    }
+
+    @Override
+    public String getType() {
+      return artifact.getType();
+    }
+
+    @Override
+    public String getClassifier() {
+      return artifact.getClassifier();
+    }
+  }
+
+  private Set<Artifact> resolveDependencies(ResolutionScope scope, boolean transitive) {
+    final Key key = ScopedProjectDependencyCache.key(project.getGroupId(), project.getArtifactId(),
+        project.getVersion(), scope, transitive);
+
+    return dependencyCache.getDependencies(key, () -> {
+      return resolveArtifacts(scope, transitive);
+    });
+  }
+
+  private Set<Artifact> resolveArtifacts(ResolutionScope scope, boolean transitive) {
+
+    DependencyResolutionResult result;
+    DependencyFilter resolutionFilter = getResolutionFilter(scope, transitive);
+
+    try {
+      result = resolveDependencies(scope, resolutionFilter);
+    } catch (DependencyResolutionException e) {
+      result = e.getResult();
+    }
+
+    Set<Artifact> artifacts = new LinkedHashSet<>();
+    if (result.getDependencyGraph() != null
+        && !result.getDependencyGraph().getChildren().isEmpty()) {
+      RepositoryUtils.toArtifacts(artifacts, result.getDependencyGraph().getChildren(),
+          Collections.singletonList(project.getArtifact().getId()), resolutionFilter);
+    }
+    return artifacts;
+  }
+
+  private DependencyFilter getResolutionFilter(ResolutionScope scope, boolean transitive) {
+    DependencyFilter filter = new ScopeDependencyFilter(null, negate(scope));
+    if (!transitive) {
+      filter = AndDependencyFilter.newInstance(filter,
+          new TransitiveDependencyFilter(project.getDependencies()));
+    }
+
+    return filter;
+  }
+
+  private static Collection<String> negate(ResolutionScope resolutionScope) {
```

Comment:
pretty sure negate is insufficiently named and javadoced.   took me a while to parse this. 
compile -> test/runtime
runtime -> system/provided/test
test -> emptySet

Add are these intentional ?
provided -> all
system -> all

Reply:
I could rename/document the method. Provided and system do not exist in io.takari.build.ResolutionScope. But I do see that I missed COMPILE_PLUS_RUNTIME, so I will add that here as well.

Reply:
maybe the method could be simplified to be clearer.
ALL.filter( s -> !allImpliedScope(resolutionScope()).contains(s))

Perhaps
notImpliedScopes() ?

Reply:
actually you are right, since the values being looped through are strings returned from ToScopes